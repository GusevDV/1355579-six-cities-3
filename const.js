export const OfferType = {
  APARTMENT: `apartment`,
  ROOM: `room`,
  HOUSE: `house`,
  HOTEL: `hotel`,
};
export const MapSetting = {
  ZOOM: 12,
  ICON_URL: `/img/pin.svg`,
  ICON_URL_ACTIVE: `/img/pin-active.svg`,
  ICON_SIZE: [30, 30]
};

export const ErrorMessage = {
  NETWROK_ERROR: `–ù–∞–º –Ω–µ —É–¥–∞–ª–æ—Å—å –ø–æ–ª—É—á–∏—Ç—å –¥–∞–Ω–Ω—ã–µ —Å —Å–µ—Ä–≤–µ—Ä–∞ üò¢`,
  COMMON_ERROR: `–ß—Ç–æ-—Ç–æ –ø–æ—à–ª–æ –Ω–µ —Ç–∞–∫. –ü–æ–ø—Ä–æ–±—É–π—Ç–µ –ø–æ–≤—Ç–æ—Ä–∏—Ç—å –ø–æ–ø—ã—Ç–∫—É üò¢`
};

export const MapDisplayType = {
  CITIES: `cities`,
  PROPERTY: `propery`,
};

export const PlaceCardType = {
  CITY: `CITY`,
  NEAR: `NEAR`,
};

export const AuthStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const Route = {
  OFFER_DETALIS: `/offer/:id`,
  LOGIN: `/login/`,
  FAVORITES: `/favorites/`
};

export const monthNames = [`January`, `February`, `March`, `April`, `May`, `June`, `July`, `August`, `September`, `October`, `November`, `December`];

export const SortIndex = {
  PRICE_LOW_TO_HIGH: 1,
  PRICE_HIGH_TO_LOW: 2,
  TOP_RATED_FIRST: 3
};
export const sortTypes = [`Popular`, `Price: low to high`, `Price: high to low`, `Top rated first`];

export const MAX_CITIES_COUNT = 6;
export const MAX_REVIEWS_COUNT = 10;
export const BASE_API_URL = `https://htmlacademy-react-3.appspot.com/six-cities`;
