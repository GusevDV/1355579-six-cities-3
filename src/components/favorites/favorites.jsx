import React from 'react';
import PropTypes from 'prop-types';
import Header from '../header/header.jsx';
import Footer from '../footer/footer.jsx';
import {connect} from 'react-redux';
import {PlaceCardType} from '../../const.js';
import PlaceCardList from '../place-card-list/place-card-list.jsx';
import {ApiCall as FavoriteOffersApiCall} from '../../reducer/favorite-offers/favorite-offers.js';
import {ApiCall as OffersApiCall} from "../../reducer/offers/offers.js";
import {getFavoriteOffersGroupByCity} from '../../reducer/favorite-offers/selectors.js';
import {offerType} from '../../types/offers-types.js';
import FavoritesEmpty from '../favorites-empty/favorites-empty.jsx';
import ThrobberSwitcher from '../throbber-switcher/throbber-switcher.jsx';

class Favorites extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.onFetchFavoriteOffers();
  }
  render() {
    const {offersGroup, isLoading, isError, onChangeFavoriteStatus} = this.props;
    return (
      <div className="page">
        <Header />
        <ThrobberSwitcher isLoading={isLoading} render={() => {
          if (Object.keys(offersGroup).length && !isError) {
            return (
              <main className="page__main page__main--favorites">
                <div className="page__favorites-container container">
                  <section className="favorites">
                    <h1 className="favorites__title">Saved listing</h1>
                    <ul className="favorites__list">
                      {Object.keys(offersGroup).map((city) => (
                        <li key={city} className="favorites__locations-items">
                          <div className="favorites__locations locations locations--current">
                            <div className="locations__item">
                              <a className="locations__item-link" href="#">
                                <span>{city}</span>
                              </a>
                            </div>
                          </div>
                          <PlaceCardList
                            offers={offersGroup[city]}
                            placeCardType={PlaceCardType.FAVORITE}
                            onFavoriteClick={onChangeFavoriteStatus}
                          />
                        </li>
                      ))}
                    </ul>
                  </section>
                </div>
              </main>
            );
          } else {
            return <FavoritesEmpty/>;
          }
        }}/>
        <Footer />
      </div>
    );
  }
}

Favorites.propTypes = {
  offersGroup: PropTypes.shape({
    [PropTypes.string]: PropTypes.arrayOf(offerType.isRequired)
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  isError: PropTypes.bool.isRequired,
  onFetchFavoriteOffers: PropTypes.func.isRequired,
  onChangeFavoriteStatus: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  offersGroup: getFavoriteOffersGroupByCity(state),
  isLoading: state.favoriteOffers.isLoading,
  isError: state.favoriteOffers.isError
});

const mapDispatchToProps = (dispatch) => ({
  onFetchFavoriteOffers: () => dispatch(FavoriteOffersApiCall.fetchFavoriteOffers()),
  onChangeFavoriteStatus: (hotelId, status) => dispatch(OffersApiCall.changeOfferFavoriteStatus(hotelId, status))
});

export {Favorites};
export default connect(mapStateToProps, mapDispatchToProps)(Favorites);


