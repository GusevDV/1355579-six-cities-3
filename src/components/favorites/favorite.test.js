import React from "react";
import renderer from "react-test-renderer";
import {BrowserRouter} from 'react-router-dom';
import {Favorites} from "./favorites.jsx";
import offersGroupMock from '../../test-mocks/favorite-group-offers.js';
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import {AuthStatus} from '../../const.js';

const mockStore = configureStore([]);

let store = mockStore({
  user: {
    authorizationStatus: AuthStatus.NO_AUTH,
    data: []
  }
});

it(`Should Favorite component render correctly when offers is loaded with success`, () => {
  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <Favorites
              offersGroup={offersGroupMock}
              isLoading={false}
              isError={false}
              onFetchFavoriteOffers={()=>{}}
              onChangeFavoriteStatus={()=>{}}
            />
          </BrowserRouter>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`Should Favorite component render correctly when offers is loaded with error`, () => {
  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <Favorites
              offersGroup={{}}
              isLoading={false}
              isError={true}
              onFetchFavoriteOffers={()=>{}}
              onChangeFavoriteStatus={()=>{}}
            />
          </BrowserRouter>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});


it(`Should Favorite component render correctly when offers is loading`, () => {
  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <Favorites
              offersGroup={{}}
              isLoading={true}
              isError={false}
              onFetchFavoriteOffers={()=>{}}
              onChangeFavoriteStatus={()=>{}}
            />
          </BrowserRouter>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
