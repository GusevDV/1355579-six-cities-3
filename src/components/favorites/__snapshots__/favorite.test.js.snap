// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Favorite component render correctly when offers is loaded with error 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="/img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/login/"
                onClick={[Function]}
              >
                <span
                  className="header__user-name user__name"
                >
                  Sign In
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <div
    className="page page--favorites-empty"
  >
    <main
      className="page__main page__main--favorites page__main--favorites-empty"
    >
      <div
        className="page__favorites-container container"
      >
        <section
          className="favorites favorites--empty"
        >
          <h1
            className="visually-hidden"
          >
            Favorites (empty)
          </h1>
          <div
            className="favorites__status-wrapper"
          >
            <b
              className="favorites__status"
            >
              Nothing yet saved.
            </b>
            <p
              className="favorites__status-description"
            >
              Save properties to narrow down search or plan yor future trips.
            </p>
          </div>
        </section>
      </div>
    </main>
  </div>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="/img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;

exports[`Should Favorite component render correctly when offers is loaded with success 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="/img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/login/"
                onClick={[Function]}
              >
                <span
                  className="header__user-name user__name"
                >
                  Sign In
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Hamburg
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places places__list tabs__content"
            >
              <a
                className="favorites__card place-card"
                href="/offer/3"
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
              >
                <div
                  className="favorites__image-wrapper place-card__image-wrapper"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/19.jpg"
                    width="260"
                  />
                </div>
                <div
                  className="place-card__info favorites__card-info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        300
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / 
                        night
                      </span>
                    </div>
                    <button
                      className="button place-card__bookmark-button--active"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <span>
                      Canal View Prinsengracht
                    </span>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                    apartment
                  </p>
                </div>
              </a>
              <a
                className="favorites__card place-card"
                href="/offer/1"
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
              >
                <div
                  className="favorites__image-wrapper place-card__image-wrapper"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/16.jpg"
                    width="260"
                  />
                </div>
                <div
                  className="place-card__info favorites__card-info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        501
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / 
                        night
                      </span>
                    </div>
                    <button
                      className="button place-card__bookmark-button--active"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "60%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <span>
                      Nice, cozy, warm big bed apartment
                    </span>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                    house
                  </p>
                </div>
              </a>
            </div>
          </li>
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Amsterdam
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places places__list tabs__content"
            >
              <a
                className="favorites__card place-card"
                href="/offer/3"
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
              >
                <div
                  className="favorites__image-wrapper place-card__image-wrapper"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="200"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/19.jpg"
                    width="260"
                  />
                </div>
                <div
                  className="place-card__info favorites__card-info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        300
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / 
                        night
                      </span>
                    </div>
                    <button
                      className="button place-card__bookmark-button--active"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <span>
                      Canal
                    </span>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                    apartment
                  </p>
                </div>
              </a>
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="/img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;

exports[`Should Favorite component render correctly when offers is loading 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="/img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/login/"
                onClick={[Function]}
              >
                <span
                  className="header__user-name user__name"
                >
                  Sign In
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <div
    className="throbber "
  >
    <img
      alt="throbber"
      className="throbber__img"
      height={100}
      src="/img/throbber.svg"
      width={100}
    />
  </div>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="/img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
